server/apis/v1_1/handler.go

```diff
@@ -813,32 +860,60 @@ func getPipelineStatus(pipeline *dfv1.Pipeline) (string, error) {
 	return retStatus, nil
 }
 
-// validatePipelineSpec is used to validate the pipeline spec
-func validatePipelineSpec(h *handler, pipeline *dfv1.Pipeline) error {
-	ns := pipeline.Namespace
+// validatePipelineSpec is used to validate the pipeline spec during create and update
+func validatePipelineSpec(h *handler, oldPipeline *dfv1.Pipeline, newPipeline *dfv1.Pipeline, validType string) error {
+	ns := newPipeline.Namespace
 	pipeClient := h.numaflowClient.Pipelines(ns)
-	valid := validator.NewPipelineValidator(h.kubeClient, pipeClient, nil, pipeline)
-	resp := valid.ValidateCreate(context.Background())
+	isbClient := h.numaflowClient.InterStepBufferServices(ns)
+	valid := validator.NewPipelineValidator(h.kubeClient, pipeClient, isbClient, oldPipeline, newPipeline)
+	var resp *admissionv1.AdmissionResponse
+	switch validType {
+	case ValidTypeCreate:
+		resp = valid.ValidateCreate(context.Background())
+	case ValidTypeUpdate:
+		resp = valid.ValidateUpdate(context.Background())
+	}
 	if !resp.Allowed {
-		errMsg := fmt.Errorf("%v", resp.Result.Message)
+		errMsg := fmt.Errorf("%s", resp.Result.Message)
 		return errMsg
 	}
 	return nil
 }
 
 // validateISBSpec is used to validate the ISB service spec
-func validateISBSpec(h *handler, isb *dfv1.InterStepBufferService) error {
-	ns := isb.Namespace
+func validateISBSpec(h *handler, prevSpec *dfv1.InterStepBufferService,
+	newSpec *dfv1.InterStepBufferService, validType string) error {
+	ns := newSpec.Namespace
 	isbClient := h.numaflowClient.InterStepBufferServices(ns)
-	valid := validator.NewISBServiceValidator(h.kubeClient, isbClient, nil, isb)
-	resp := valid.ValidateCreate(context.Background())
+	valid := validator.NewISBServiceValidator(h.kubeClient, isbClient, prevSpec, newSpec)
+	var resp *admissionv1.AdmissionResponse
+	switch validType {
+	case ValidTypeCreate:
+		resp = valid.ValidateCreate(context.Background())
+	case ValidTypeUpdate:
+		resp = valid.ValidateUpdate(context.Background())
+	}
 	if !resp.Allowed {
-		errMsg := fmt.Errorf("%v", resp.Result.Message)
+		errMsg := fmt.Errorf("%s", resp.Result.Message)
 		return errMsg
 	}
 	return nil
 }
 
+// validateNamespace is used to validate the namespace for a pipeline spec
+// For a request, the namespace provided as parameter should be same as the namespace in the pipeline spec
+// if the namespace is not provided in the pipeline spec, the default namespace will be used
+func validateNamespace(h *handler, pipeline *dfv1.Pipeline, ns string) error {
+	var errMsg error
+	errMsg = nil
+	if pipeline.Namespace == "" && ns != DefaultNamespace {
```

Comment:
What is the reason we need to check DefaultNamespace?

Reply:
If the namespace is not provided in the spec, then we would need to compare with default namespace right?

Reply:
Okay my bad, I realised we do not need to compare with default namespace
If no namespace in spec then we should just use the one from the URL